### PR TITLE
storagenode/trust: implement fetching peer identity without kademlia and endpoint

### PR DESCRIPTION
**Pull request by egonelbre**
Ensure that we directly dial the whitelisted satellites instead of going through kademlia.

To implement, it sets up a custom credentials which captures the auth info during the initial handshake, this avoids the necessity for a separate Endpoint for fetching information.

As a consequence it also removes ability to trust all satellites, which was a bad idea in the first place.

Why: Currently kademlia lookups keep failing, however it's much faster and reliable to dial directly.

Please describe the tests:
 - Test 1:
 - Test 2:
 
Please describe the performance impact:

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Is the PR appropriately sized? (If it could be broken into smaller PRs it should be)
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?
 - [ ] Do the database access patterns make sense?

### CODE REVIEW

File: storagenode/trust/service.go [hunk 1]
```diff
@@ -23,55 +24,39 @@ var mon = monkit.Package()
...
 
 // satelliteInfoCache caches identity information about a satellite
 type satelliteInfoCache struct {
 	mu       sync.Mutex
+	url      storj.NodeURL
 	identity *identity.PeerIdentity
-	nodeURL  storj.NodeURL
 }
 
 // NewPool creates a new trust pool using kademlia to find certificates and with the specified list of trusted satellites.
```
Comment (ifraixedes) [inline storagenode/trust/service.go]: Update this doc comment

File: storagenode/trust/service.go [hunk 2]
```diff
@@ -93,29 +78,15 @@ func (pool *Pool) GetSignee(ctx context.Context, id storj.NodeID) (_ signing.Sig
...
-				pool.trustedSatellites[id] = info
-			}
-			pool.mu.Unlock()
-		}
-	} else {
-		if !ok {
-			return nil, fmt.Errorf("signee %q is untrusted", id)
-		}
+	if !ok {
+		return nil, fmt.Errorf("signee %q is untrusted", id)
```
Comment (ifraixedes) [inline storagenode/trust/service.go]: should we have a class for this error? (it's probably to do it in another PR)

File: pkg/transport/fetchidentity.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	AuthInfo credentials.AuthInfo
+}
+
+// ClientHandshake does the authentication handshake specified by the corresponding
+// authentication protocol on rawConn for clients. It returns the authenticated
+// connection and the corresponding auth information about the connection.
+func (capture *handshakeCapture) ClientHandshake(ctx context.Context, s string, conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ClientHandshake(ctx, s, conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
```
Comment (navillasa) [inline pkg/transport/fetchidentity.go]: nit - does the error need to be wrapped?

File: pkg/transport/fetchidentity.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// ServerHandshake does the authentication handshake for servers. It returns
+// the authenticated connection and the corresponding auth information about
+// the connection.
+func (capture *handshakeCapture) ServerHandshake(conn net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	conn, auth, err := capture.TransportCredentials.ServerHandshake(conn)
+	capture.AuthInfo = auth
+	return conn, auth, err
```
Comment (navillasa) [inline pkg/transport/fetchidentity.go]: nit - does the error need to be wrapped?

File: pkg/transport/fetchidentity.go (hunk 3 above)
Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: here, no, because it's wrapping and we don't know how the error is handled internally, by wrapping we might lose important interface methods.

File: pkg/transport/fetchidentity.go (hunk 4 above)
Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: here, no, because it's wrapping and we don't know how the error is handled internally, by wrapping we might lose important interface methods.

File: pkg/transport/fetchidentity.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	options := append([]grpc.DialOption{
+		grpc.WithTransportCredentials(capture),
+		grpc.WithBlock(),
+		grpc.FailOnNonTempDialError(true),
+		grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
+			conn, err := (&net.Dialer{}).DialContext(ctx, "tcp", addr)
+			if err != nil {
+				return nil, err
```
Comment (navillasa) [inline pkg/transport/fetchidentity.go]: nit - error wrap?

File: pkg/transport/fetchidentity.go (hunk 4 above)
Comment (ifraixedes) [reply pkg/transport/fetchidentity.go]: We have to find out a way to let the callers know with errors (mostly methods that return errors) are fine to be inspected like this one and which ones not.

File: pkg/transport/fetchidentity.go (hunk 5 above)
Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: no, because we are here, no, because it is wrapping internal behavior of grpc and we don't know how the error is handled internally, by wrapping we might lose important interface methods.

Comment (ifraixedes) [reply pkg/transport/fetchidentity.go]: this one bubbles up in the call `conn, err := grpc.DialContext(timedCtx, node.GetAddress().Address, options...)`, no?

File: storagenode/trust/service.go (hunk 2 above)
Comment (navillasa) [reply storagenode/trust/service.go]: agree, I think this could be its own err class vs fmt. but I think that would be fine to just add at the top of this file in this pr like `var ErrUntrusted = errs.Class("signee is untrusted:")` so then you could do `return nil, ErrUntrusted.New(id.String())`. but also I don't think this is a critical change.

Comment (egonelbre) [reply storagenode/trust/service.go]: I currently made it use `Error` directly, since we don't need the distinction at the moment, I didn't add a new class. When we do, we can add more.

File: pkg/transport/fetchidentity.go (hunk 5 above)
Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: Not sure what you mean?

File: pkg/transport/fetchidentity.go (hunk 4 above)
Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: I suspect that the error warpping might hide essential interfaces, such as `interface { Temporary() bool }`. https://godoc.org/google.golang.org/grpc#FailOnNonTempDialError, however I haven't checked the errs implementation for it.

File: pkg/transport/fetchidentity.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/peertls"
+)
+
+// TODO: check whether this would be problematic with concurrency or should we thread the authinfo through context?
```
Comment (zeebo) [inline pkg/transport/fetchidentity.go]: I'm worried about this TODO. I don't want data races causing weird memory corruption because `AuthInfo` is an interface and so you can tear the writes. It's not too hard to just add a mutex.

Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: Yeah, it's less about concurrent writes, I'm not sure how grpc internally calls the handshakes, i.e. does it call it once, in parallel, does it cache or something else.

Comment (egonelbre) [reply pkg/transport/fetchidentity.go]: Either way, I'll add a mutex.

File: pkg/transport/fetchidentity.go (hunk 4 above)
Comment (ifraixedes) [reply pkg/transport/fetchidentity.go]: sure, I was only saying that we should find away that the callers can know which methods return original errors so they can do type assertions.

File: pkg/transport/fetchidentity.go (hunk 5 above)
Comment (ifraixedes) [reply pkg/transport/fetchidentity.go]: When this function (which is an option) is called and due that its result is returned?

I think that the function gets called by `grpc.DialContext(...` as its receiving the options.
